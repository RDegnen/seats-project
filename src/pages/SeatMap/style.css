.selectedSeatRow {
  display: flex;
  justify-content: center;
}

.selectedSeatCol {
  margin: 10px 20px;
  font-weight: bold;
  color: grey;
  font-size: 30px;
}