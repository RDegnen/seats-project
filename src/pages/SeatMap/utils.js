// Group seats into objects based off seat properties
export const groupBy = (seats, prop) => {
  const returnObj = {}
  seats.forEach(seat => {
    if (returnObj[seat[prop]]) {
      returnObj[seat[prop]].push(seat)
    } else {
      returnObj[seat[prop]] = [seat]
    }
  })
  return returnObj
}
// Sort the seats in a row
export const sortRows = rows => {
  const sortedSeats = {...rows}
  for (const key in sortedSeats) {
    sortedSeats[key].sort((a, b) => a.seat > b.seat)
  }
  return sortedSeats
}
// Add ailes to all rows
export const addAisles = planeClass => {
  const alpha = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
  Object.keys(planeClass).forEach(key => {
    const row = planeClass[key]
    for (let i = 0; i < row.length; i++) {
      if (row[i].seat !== alpha.charAt(i)) {
        row.splice(i, 0, { 
          row: parseInt(key),
          seat: alpha.charAt(i),
          aisle: true 
        })
      }
    }
  })
}
// Create the seat layout 
export const createLayout = res => {
  const classes = groupBy(res, 'class')
  const seats = {}
  for (const key in classes) {
    seats[key] = sortRows(groupBy(classes[key], 'row'))
  }
  Object.keys(seats).map(key => addAisles(seats[key]))
  return seats
}