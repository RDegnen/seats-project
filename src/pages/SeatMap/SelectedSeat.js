import React from 'react'
import PropTypes from 'prop-types'
import { Row, Col } from 'reactstrap'

import './style.css'

const SelectedSeat = ({ seat, columns }) => (
  <Row className="selectedSeatRow">
    {columns.map(({ title, value }) => (
      <Col key={title} className="selectedSeatCol" md="3">
        {title}: {value(seat)}
      </Col>
    ))}
  </Row>
)

SelectedSeat.propTypes = {
  seat: PropTypes.object,
  columns: PropTypes.array,
}

export default SelectedSeat