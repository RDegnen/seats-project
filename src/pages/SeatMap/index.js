import React, { useEffect, useState } from 'react'

import SeatGrid from '../../components/SeatGrid'
import { createLayout } from './utils'

const seatId = (seatClass, row, seat) => `${seatClass}-${row}-${seat}`

const SeatMap = () => {
  const [seats, setSeats] = useState({})
  const [selectedSeat, setSelected] = useState(null)

  const getSeats = async () => {
    const res = 
      await fetch('https://s3.amazonaws.com/frontend-candidate-homework.lola.co/seats.json', {
          method: 'GET',
          mode: 'cors',
        }).then(res => res.json())
          .then(json => json)
          .catch(err => err)

    setSeats(createLayout(res))
  }

  useEffect(() => {
    getSeats()
  }, [])

  const selectSeat = item => {
    const { occupied, row, seat, class: seatClass } = item
    if (occupied) {
      return
    } else if (selectedSeat) {
      selectedSeat.selected = false
      if (seatId(
            selectedSeat.class, 
            selectedSeat.row, 
            selectedSeat.seat
          ) === seatId(seatClass, row, seat)) {
        setSelected(null)
        return
      }
    }
    item.selected = true
    setSelected(item)
  }

  return (
    <SeatGrid seats={seats} selectSeat={selectSeat}/>
  )
}

export default SeatMap