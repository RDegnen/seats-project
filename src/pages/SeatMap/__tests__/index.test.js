import React from 'react'
import { shallow } from 'enzyme'

import SeatMap from '../'

global.fetch = jest.fn(() => new Promise(resolve => resolve()));

const getComponent = props => shallow(<SeatMap {...props} />)

describe('<SeatMap />', () => {
  it('should match snapshot', () => {
    const component = getComponent()
    expect(component).toMatchSnapshot()
  })
})