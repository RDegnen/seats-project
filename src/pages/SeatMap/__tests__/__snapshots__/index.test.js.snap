// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SeatMap /> should match snapshot 1`] = `
<SeatGrid
  seats={Object {}}
  selectSeat={[Function]}
/>
`;
